src/timeout.py

```diff
@@ -20,96 +20,164 @@
-        for child in child_nodes:
-            if child['state'] != 'done':
-                child = self._set_node_result(child)
-                child['state'] = 'done'
-                self._db.submit({'node': child})
-                self._update_child_nodes(child['_id'])
-
-    def _update_timed_out_node(self, node):
-        """Set Node state to done if maximum wait time is over"""
-        if node['state'] != 'done':
-            current_time = datetime.utcnow()
-            expires = datetime.fromisoformat(node['timeout'])
-            if current_time > expires:
-                self._logger.log_message(logging.INFO,
-                                         f"Node timed-out {node['_id']}")
-                self._update_child_nodes(node['_id'])
-                node = self._set_node_result(node)
-                node['state'] = 'done'
-                self._db.submit({'node': node})
-            elif node['state'] == 'available':
-                self._update_available_node(node, current_time)
+                self.log.debug(f"{node_id} holdoff left: {holdoff - now}")
+                sleep = min(sleep, holdoff - now)
+        self._submit_lapsed_nodes(closing_nodes, 'closing', 'HOLDOFF')
+        self._submit_lapsed_nodes(timeout_nodes, 'done', 'DONE')
+        return sleep.total_seconds()
 
     def _run(self, ctx):
-        self.log.info("Looking for timed-out nodes...")
+        self.log.info("Looking for nodes with lapsed holdoff...")
         self.log.info("Press Ctrl-C to stop.")
 
         while True:
-            nodes = self._db.get_nodes()
-            for node in nodes:
-                self._update_timed_out_node(node)
+            available_nodes = self._get_available_nodes()
+            sleep_time = self._check_available_nodes(available_nodes)
+            sleep(sleep_time)
+
+        return True
+
+
+class Closing(TimeoutService):
+
+    def __init__(self, configs, args):
+        super().__init__(configs, args, 'timeout-closing')
+
+    def _get_closing_nodes(self):
+        nodes = self._db.get_nodes({'state': 'closing'})
+        return {node['_id']: node for node in nodes}
+
+    def _check_closing_nodes(self, closing_nodes):
+        done_nodes = {}
+        for node_id, node in closing_nodes.items():
+            running = self._count_running_child_nodes(node_id)
+            self.log.debug(f"{node_id} RUNNING: {running}")
+            if not running:
+                done_nodes[node_id] = node
+        self._submit_lapsed_nodes(done_nodes, 'done', 'DONE')
+
+    def _run(self, ctx):
+        self.log.info("Looking for nodes that are done closing...")
+        self.log.info("Press Ctrl-C to stop.")
+
+        while True:
+            closing_nodes = self._get_closing_nodes()
+            self._check_closing_nodes(closing_nodes)
             sleep(self._poll_period)
 
         return True
 
 
+MODES = {
+    'timeout': Timeout,
+    'holdoff': Holdoff,
+    'closing': Closing,
+}
+
+
```

Comment:
Maybe we could think of using `Enum` here, too?

Reply:
Yes, in fact I think this will need to be revisited as a follow-up as ideally we should be getting the list of states from the API code.  I'd rather leave it out of this PR for now as it's already quite big.